src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionary.cs

```diff
@@ -369,6 +370,73 @@ public void SetModelValue([NotNull] string key, [NotNull] ValueProviderResult va
             GetModelStateForKey(key).Value = value;
         }
 
+        /// <summary>
+        /// Clears <see cref="ModelStateDictionary"/> entries for model in specified <see cref="ModelMetadata"/>.
+        /// </summary>
+        /// <param name="modelMetadata">The <see cref="ModelMetadata"/>.</param>
+        /// <param name="modelKey">The entry to clear. </param>
+        /// <param name="modelMetadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        public void ClearModelStateDictionaryForModel(
+            [NotNull] ModelMetadata modelMetadata,
+            [NotNull] string modelKey,
+            IModelMetadataProvider modelMetadataProvider)
+        {
+            if (modelMetadata.IsCollectionType)
+            {
+                var elementType = ModelBindingHelper.GetElementType(modelMetadata.ModelType);
+                var elementMetadata = modelMetadataProvider.GetMetadataForType(elementType);
+
+                ClearModelStateDictionaryForModel(elementMetadata, modelKey);
+            }
+            else
+            {
+                ClearModelStateDictionaryForModel(modelMetadata, modelKey);
+            }
+        }
+
+        /// <summary>
+        /// Clears <see cref="ModelStateDictionary"/> entries for model in specified <see cref="ModelMetadata"/>.
+        /// </summary>
+        /// <param name="modelMetadata">The <see cref="ModelMetadata"/>.</param>
+        /// <param name="modelKey">The entry to clear.</param>
+        public void ClearModelStateDictionaryForModel(
+            [NotNull] ModelMetadata modelMetadata,
+            [NotNull] string modelKey)
+        {
+            foreach (var property in modelMetadata.Properties)
+            {
+                var propertyBindingName = property.BinderModelName ?? property.PropertyName;
+                var childKey = ModelBindingHelper.CreatePropertyModelName(modelKey, propertyBindingName);
+
+                ClearModelStateDictionaryEntries(childKey);
+            }
+        }
+
+        /// <summary>
+        /// Clears <see cref="ModelStateDictionary"/> entries that match the key that is passed as parameter.
+        /// </summary>
+        /// <param name="key">The key of <see cref="ModelStateDictionary"/> to clear.</param>
+        public void ClearModelStateDictionaryEntries([NotNull] string key)
+        {
+            var pattern = "^(" + key + "|(" + key + @"\[\d*\]\..)|(" + key + @"\..)|(\[\d*\]\." +
```

Comment:
What is the benefit of using a regex? It seems like doing a few `string.StartsWith` checks would achieve the same thing, no? Like:

`if (dictionaryKey.StartsWith(key) ||
    dictionaryKey.StartsWith(key + "[") ||
    dictionaryKey.StartsWith(key + ".") ||
    <whatever else we need>)
{
    ...
}
```

RegEx strikes me as overkill for this... I don't think we use regexes anywhere else in the system.